rts/motoko-rts-macros/src/lib.rs

```diff
@@ -167,3 +167,160 @@ pub fn non_incremental_gc(attr: TokenStream, input: TokenStream) -> TokenStream
+            output = quote!(
+                #output
+                #macro_ident!(#length_syn; #parts);
+            );
+        }
+    }
+
+    // return syn::Error::new_spanned(quote!(), format!("Macro expansion:\n{}", output))
+    //     .to_compile_error()
+    //     .into();
+
+    output.into()
+}
+
+/// This macro wraps `#[ic_mem_fn]` with automatic type conversions for
+/// Motoko types using traits defined in the `motoko_rts::custom` module.
+#[proc_macro_attribute]
+pub fn motoko(attr: TokenStream, input: TokenStream) -> TokenStream {
+    let ic_mem_attr: proc_macro2::TokenStream = attr.into();
+
+    // Wrapped function
+    let mut fun = syn::parse_macro_input!(input as syn::ItemFn);
+    let fn_ident = &fun.sig.ident;
+    let fn_name = fn_ident.to_string();
+    let fn_params: Vec<(ArgSort, syn::Ident, syn::Type)> = fun
+        .sig
+        .inputs
+        .iter_mut()
+        .enumerate()
+        .map(|(i, arg)| match arg {
+            syn::FnArg::Receiver(_) => {
+                // TODO: replace panic with macro error message
+                panic!("IC functions can't have receivers (`&self`, `&mut self`, etc.)")
+            }
+            syn::FnArg::Typed(pat) => {
+                let mut sort = ArgSort::Value;
+                pat.attrs.retain(|attr| {
+                    if attr.meta.to_token_stream().to_string() == "memory" {
+                        sort = ArgSort::Memory;
+                        false
+                    } else {
+                        true
+                    }
+                });
+                (
+                    sort,
+                    syn::Ident::new(&format!("arg{}", i), proc_macro2::Span::call_site()),
+                    (*pat.ty).clone(),
+                )
+            }
+        })
+        .collect();
+
+    // Motoko parameters
+    let params: Vec<proc_macro2::TokenStream> = fn_params
+        .iter()
+        .filter_map(|(sort, ident, _ty)| match sort {
+            ArgSort::Value => Some(quote!(#ident: crate::types::Value)),
+            ArgSort::Memory => None,
+        })
+        .collect();
+
+    // Memory parameter
+    let memory_ident = syn::Ident::new("memory", proc_macro2::Span::call_site());
+
+    // Motoko -> Rust conversions
+    let arg_conversions: Vec<proc_macro2::TokenStream> = fn_params
+        .iter()
+        .filter_map(|(sort, ident, _)| match sort {
+            ArgSort::Value => Some(quote!(
+                let #ident = crate::custom::FromValue::from_value(#ident, #memory_ident).unwrap()
+            )),
+            ArgSort::Memory => None,
+        })
+        .collect();
+
+    // Motoko args
+    let args: Vec<proc_macro2::TokenStream> = fn_params
+        .iter()
+        .map(|(sort, ident, _)| match sort {
+            ArgSort::Value => quote!(#ident),
+            ArgSort::Memory => quote!(#memory_ident),
+        })
+        .collect();
+
+    // Motoko return value
+    let ret = quote!(crate::types::Value);
```

Comment:
I am not sure whether all FFI functions would have a `Value` return type. Maybe there are some functions without return type (e.g. that implement checks with traps or do some updates).